output/pixel_pusher.c

```diff
@@ -0,0 +1,272 @@
+#include "pixel_pusher.h"
+
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <SDL2/SDL.h>
+#include <sys/socket.h>
+#include <sys/time.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "output/config.h"
+#include "util/err.h"
+
+// This file implements a PixelPusher output.  It is designed to discover only 1 PixelPusher,
+// but to use multiple grids attached to that PixelPusher
+
+static struct pp_discovery_packet * pp_info = NULL;
+
+static struct pp_device * grid_devices = NULL;
+static size_t n_grid_devices = 0;
+
+// Reusable state for sending data packets
+static int out_fd = -1;
+static uint32_t seq_num = 0;
+static struct sockaddr_in * out_addr = NULL;
+
+static uint8_t * out_packet;
+
+// See output_pp_do_frame for why we need this
+static struct timespec packet_interval = { .tv_sec = 0, .tv_nsec = 500*1000};
+
+static int pp_find_pusher() {
+    INFO("Initializing PixelPusher");
+
+    // Try to open a socket
+    int server_fd = socket(AF_INET, SOCK_DGRAM, 0);
+    if (server_fd < 0) {
+        PERROR("Error opening PixelPusher discovery socket");
+        return -1;
+    }
+
+    // Set its timeout
+    struct timeval tv;
+    tv.tv_sec = output_config.pixel_pusher.discovery_seconds;
+    if (!tv.tv_sec) {
+        ERROR("PixelPusher discovery seconds must be non-zero");
+        close(server_fd);
+        return -1;
+    }
+
+    tv.tv_usec = 0;
+    if (setsockopt(server_fd, SOL_SOCKET, SO_RCVTIMEO, (char *) &tv, sizeof(struct timeval)) < 0) {
+        PERROR("Error setting PixelPusher discovery socket timeout");
+        close(server_fd);
+        return -1;
+    }
+
+    // Socket address
+    struct sockaddr_in server_addr;
+    server_addr.sin_family = AF_INET;
+    server_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+    server_addr.sin_port = htons(output_config.pixel_pusher.port);
+
+    // Bind the socket
+    if (bind(server_fd, (struct sockaddr *) &server_addr, sizeof server_addr) < 0) {
+        PERROR("Error binding PixelPusher discovery socket");
+        close(server_fd);
+        return -1;
+    }
+
+    // Read
+    if (pp_info) {
+        free(pp_info);
+    }
+    pp_info = calloc(1, sizeof *pp_info);
+    if (pp_info == NULL) MEMFAIL();
+
+    size_t expected_bytes = sizeof *pp_info;
+    int bytes_read = read(server_fd, pp_info, expected_bytes);
+    if (bytes_read < (int) expected_bytes) {
+        PERROR("Expected to read %zu bytes from PixelPusher but read %d bytes", expected_bytes, bytes_read);
+        free(pp_info);
+        pp_info = NULL;
+        close(server_fd);
+        return -1;
+    }
+
+    close(server_fd);
+
+    struct in_addr ip_addr;
```

Comment:
Using `inet_ntoa` seems to be the standard way to print out an IP address in network order; I think it's a pretty helpful info message so I don't feel bad about using a couple extra bytes of stack space during an initialization function that's only run once.